designdocs/recovery_system.md

```diff
@@ -0,0 +1,505 @@
+
+### Key Generation (Email + Salt/Seed)
+
+The blockchain combines the user’s email with a random seed in a reversible but secure way.
+Two design options were considered:
+
+- Option A: Deterministic Hash (One-Way): Compute a cryptographic hash H = SHA-256(email || seed || globalSalt) and store H on-chain. Here || denotes concatenation and globalSalt could be a chain-specific constant or omitted if seed is sufficiently random. In this scheme, reversibility means that given the correct email and seed, one can recompute H and match the on-chain value – thus verifying the combination. This is effectively a commitment to the email + seed. The seed itself (the user’s recovery code) is not stored on-chain, only the hash, so it cannot be directly recovered by anyone. The hash approach is simple: any recovery provider can later hash the user’s provided email + seed and check against the on-chain record for a match.
+
+- Option B: Encryption (Two-Way): Derive an encryption key from the email and seed, and use it to encrypt a known plaintext (such as the user’s MSA ID or a fixed marker). For example, derive a symmetric key K = PBKDF2(email || seed, salt=networkSalt, iter=100k) and use AES-256 to encrypt the user’s MSA ID. Store the ciphertext on-chain. In the future, a recovery provider who is given the correct email and seed can derive K and decrypt the blob to retrieve the MSA ID (achieving “reversibility” by decryption). This approach hides the MSA ID and requires possession of the seed to decrypt. It effectively binds the MSA ID to the secret.
+
+Option A (hash-based) is favored for simplicity and on-chain verification capability. Hashing the recovery data means the chain can later verify a recovery attempt by recomputing a hash, without needing to store any encryption keys. Hash storage is also aligned with security best practices. The only data needed for recovery is a successful hash match, rather than extracting plaintext.
+
+### Hash Algorithm and Salting
+
+This design uses a modern hash or Key Derivation Function (KDF) to protect the recovery code. A good choice is Argon2id or PBKDF2-HMAC-SHA256 with a high iteration count, as recommended by OWASP. For instance, Argon2id with ~ (2^16) memory and 3 iterations, or PBKDF2 with 200,000+ iterations, will slow down any brute-force attempt substantially. In our design, the “salt” for the hash can be the user’s email or MSA ID (to ensure uniqueness per user) combined with a global constant. However, since the seed itself is random and unique, an explicit additional salt is optional – Argon2/PBKDF2 internally use a salt. The main point is to avoid storing a raw unsalted hash of a low-entropy value. Here the entropy comes mostly from the seed phrase, which is very high.
```

Comment:
I don't think a "global salt" adds any security since it would be known by any attacker; it still comes down to the guessability of the identifier and seed phrase.